### PR TITLE
Work in progress: Fixing the colors.

### PR DESCRIPTION
**Pull request by andyleejordan**
Please see the commit messages. I'm not sure how we got to this state, maybe Windows changed, but it seems that the notion of "dark" flipped on us. The original color scheme broke down in the Bash on Ubuntu on Windows terminal, as all the colors were mapping to their dark versions, which are used as the Solarized background tones.

This is a work-in-progress, as I've only fixed it for the Dark color-scheme so far. I'll add screenshots of the before and after, and comparison to my MacBook terminal, but I wanted to share this early to see if anyone knows _why_ our colors were so off.

### CODE REVIEW

**Comment by rbeesley** [comment]
That was something I noticed early on, it was as though the intensity bit was interrupted to mean the opposite. This is shared amongst other implementations like PuTTY.

There's a little more at work here, too. As I recall, the terminal defines 20 colors while the console has only 16. Depending on what is selected, you may or may not see the mapping you expect or desire.

Hence all of this work was a compromise to achieve the desired results. If you can come up with a configuration which has better support for cmd.exe, PowerShell, and bash.exe, I'd absolutely welcome the contribution. We just need to make sure that it plays well with all those environments, and that has been the struggle. All of my efforts usually broke something in an undesirable way.

**Comment by andschwa** [comment]
> All of my efforts usually broke something in an undesirable way.

Oh yeah, I was having this problem. So far, I have Bash on Ubuntu on Windows working nicely, including when SSH'd into an actual Linux machine.

Using Fish to test the 16 colors:

```
function show_colors
    set_color -c | xargs -I'{}' fish -c "set_color {}; echo '{}'; set_color normal"
end
```

Bash on Ubuntu on Linux SSH'd to a real Ubuntu machine (with `xterm-256color` blah blah blah):

wsl

Same thing but on the actual WSL using Fish, no SSH'ing:
wsl_nossh

And as I took a screenshot of PowerShell, I noticed a bug... fixing.

**Comment by andschwa** [comment]
I confused myself, the PowerShell colors are as "right" as can be, considering there isn't a 1-1 mapping. But I've added the descriptions so we don't get confused later.

powershell

**Comment by andschwa** [comment]
And this is what I'm comparing against on my MacBook with iTerm in Solarized theme:

iterm colors

**Comment by andschwa** [comment]
I've so far been really happy with how this is working; haven't run into anything it breaks. If I applied the same fixes to the light variants, would you be interested in accepting a proper PR?

**Comment by rbeesley** [comment]
I think I would have expected more to change in Set-SolarizedDarkColorDefaults.ps1. Is there a reason why some of the colors were left in their original states while the other were changed?

File: solarized-dark.reg [hunk 1]
```diff
@@ -6,42 +6,53 @@ Windows Registry Editor Version 5.00
...
+"ColorTable08"=dword:00362b00
+"ColorTable09"=dword:00969483
+"ColorTable10"=dword:00756e58
+"ColorTable11"=dword:00a1a193
+"ColorTable12"=dword:00164bcb
+"ColorTable13"=dword:00c4716c
+"ColorTable14"=dword:00837b65
+"ColorTable15"=dword:00d5e8ee
+
+; Note, the ScreenColors and PopupColors map the screen/popup
```
Comment (rbeesley) [inline solarized-dark.reg:48]: Do you have a reference to this somewhere? I believe it is incorrect. I would expect `"ScreenColors"=dword:00000089`. Look at how I previously had PopupColors. That referred to ColorTable entries F/6.

**Comment by andschwa** [comment]
@rbeesley The reason is that the actual colors are changed by the registry keys. The PS1 file just changes which of those colors is used for particular fonts. Because we swapped the Dark and Non-Dark variants of colors in the registry file, those were the only ones we needed to swap in the PS1 (and fixed the background color to be the Solarized-specified background color of "Black" AKA base03). So for example, `DarkCyan` as a color in PS1 intended by the Solarized Cyan color, which with the reg fixes, is correctly `Cyan` now, as `DarkCyan` is base1 (and so on and so forth).

**Comment by andyleejordan** [comment]
@rbeesley Should I fix the light variant too and open a PR?

**Comment by andyleejordan** [comment]
@rbeesley Well that's confusing; it looks like Windows reversed the change, and now `master` is working correctly (at least for the default command prompt). I wonder if they had a regression and fixed it.

**Comment by andyleejordan** [comment]
Spoke too soon; the accent colors are still incorrect on `master`, but the background is correct (and now incorrect in my branch).

**Comment by andyleejordan** [comment]
After experimenting, I think there is _currently_ a regression. `ScreenColors` is supposed to be of the form `0008 0009` where the first four bytes are the background color and the last four bytes are the text color. This is what we previously documented, and indeed, it's how `PopupColors` is _still_ behaving.

However, it appears that the last two bytes of `ScreenColors` are being read for the colors, instead of two pairs of four bytes. That is, what I've set is being interpreted as `09` ending up as `base02` (color 0) and `base00` (color 9). If I edit the default value in the GUI of cmd.exe to use `base03` and `base00` (colors 8 and 9), and save this, the registry key `ScreenColors` is then saved as `0000 0089`.

So instead of the first four bytes being the background color, it appears to me that a console regression is instead reading the second to last byte as the background color.

Windows Version 1803 OS Build 17134.228

/cc @bitcrazed

**Comment by rbeesley** [comment]
Goodness. If you have something which is working, go ahead and contribute a change. @neilpa is the only one with the ability to commit code I believe. However, it sounds like maybe you should open a bug against the console too. I know the console team has been doing a lot of work, just read their blog posts. It's possible there has been a breaking change, but I haven't spent a lot of time on this side project since so much is changing in Windows 10. They built a colortool utility which changes the console defaults, so I'm wondering if there's a lot of return on this investment any more. If users are using a recent Windows 10 build, it's probably easier just going that way.

**Comment by andyleejordan** [comment]
I am following up internally with the console team PM on the odd behavior of `ScreenColors`. This PR works right now (besides `ScreenColors`, which I can fix but then would break if that is indeed a regression), but doesn't do the light variant.

I'll wait and figure out what's up with `ScreenColors`, and find out if the team has a better way to "skin" the terminal, and if not, I'll do the light version, update the PR, and bug the maintainer 😄 

I gotta have my Solarized!

**Comment by neilpa** [comment]
>  @neilpa is the only one with the ability to commit code I believe.

@rbeesley I _think_ you have the ability to merge PRs (I thought that's what contributor status did). I'll double check the settings because at this point it's been years since I've been on a Windows machine.

**Comment by neilpa** [comment]
I was confusing contributor and collaborator. @rbeesley once you accept the invite you should be good. I'll leave the merge to you since I'm pretty far removed at this point.

**Comment by andyleejordan** [comment]
@rbeesley The Windows guys set me straight, and I've updated the PR to work correctly.

Now, do we want to do the light variant in this PR as well?

**Comment by rbeesley** [comment]
@andschwa, I'm curious what they told you. If this is the right fix for Dark, then we should probably make a similar fix for Light.

**Comment by andyleejordan** [comment]
@rbeesley It is right here https://github.com/Microsoft/console/issues/267

My understanding of `ScreenColors` was incorrect; and we'd been getting lucky because an index was `0`, which happened to be in both the second lowest four bits, and the fifth lowest.

I will get Light fixed up, and re-tested, with test screenshots. Might be a little bit but it'll happen!

**Comment by rbeesley** [comment]
I'd apparently had that review comment uncommitted since Mar 13... I guess you never saw it. Seems to align with what you heard back.

**Comment by andyleejordan** [comment]
Bah haha, latest Windows update updated PSReadLine and changed the property names of the colors... more to fix! Instead of XForegroundColor and XBackgroundColor, it's now just XColor. It's not backwards compatible either.

**Comment by andyleejordan** [comment]
I have no idea if this is relevant any more...and TBH I'm now using the modern Windows Terminal (the beta) with WSL2, and rarely touch PowerShell or the command prompt, even on Windows.

**Pull request closed without merging**